follower/execution/follower.go

```diff
@@ -0,0 +1,299 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package execution
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/dgraph-io/badger/v2"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/optakt/flow-dps/models/dps"
+)
+
+const cacheSize = 1024
+
+type blockReader interface {
+	Notify() <-chan string
+	Read(filename string) ([]byte, error)
+}
+
+type Follower struct {
+	log zerolog.Logger
+
+	blocks blockReader
+	codec  dps.Codec
+	db     *badger.DB
+
+	data    map[uint64]*BlockData
+	current *BlockData
+	height  uint64
+	index   int
+
+	stop chan struct{}
+}
+
+func New(log zerolog.Logger, blocks blockReader, codec dps.Codec, db *badger.DB) *Follower {
+	f := Follower{
+		log: log,
+
+		blocks: blocks,
+		codec:  codec,
+		db:     db,
+
+		height: math.MaxUint64,
+		data:   make(map[uint64]*BlockData, cacheSize),
+
+		stop: make(chan struct{}),
+	}
+
+	return &f
+}
+
+// Run launches the execution follower.
+func (f *Follower) Run() error {
+	tick := time.NewTicker(1 * time.Second)
+	defer tick.Stop()
+
+	for range tick.C {
+		select {
+		case <-f.stop:
+			return nil
+		default:
+		}
+
+		// If cache is already full, wait for items to get consumed.
+		if len(f.data) == cacheSize {
+			continue
+		}
+
+		// If cache has available space, wait for new block data to be available.
+		select {
+		case <-f.stop:
+			return nil
+		case file := <-f.blocks.Notify():
+			b, err := f.blocks.Read(file)
+			if err != nil {
+				return fmt.Errorf("could not read block data: %w", err)
+			}
+
+			var data BlockData
+			err = f.codec.Unmarshal(b, &data)
+			if err != nil {
+				return fmt.Errorf("could not decode block data: %w", err)
+			}
+			if data.Block == nil {
+				return fmt.Errorf("invalid block data")
+			}
+
+			f.data[data.Block.Header.Height] = &data
+
+			// This should only ever happen once, when we get the oldest available block.
+			if data.Block.Header.Height < f.height {
+				if f.height != math.MaxUint64 {
+					// This should never happen, it would mean we received blocks in an
+					// incorrect order!
+					return fmt.Errorf("fatal discrepancy: received block data in incorrect order")
+				}
+				f.height = data.Block.Header.Height
+			}
```

Comment:
That seems like a bunch of code that could be eliminated with more elegant logic.